### PR TITLE
Add batched federation entity resolvers

### PR DESCRIPTION
**Pull request by spawnia**
- [x] Added or updated tests
- [x] Documented user facing changes
- [x] Updated CHANGELOG.md

Resolves https://github.com/nuwave/lighthouse/issues/1996

**Changes**

**Breaking changes**

None

### CODE REVIEW

File: src/Federation/Resolvers/Entities.php [hunk 1]
```diff
@@ -32,11 +33,25 @@ public function __invoke($root, array $args, GraphQLContext $context, ResolveInf
...
+            $resolver = $this->entityResolverProvider->resolver($typename);
+            if ($resolver instanceof BatchedEntityResolver) {
+                foreach ($resolver($representations) as $result) {
+                    $results[] = $result;
+                }
+            } else {
+                foreach ($representations as $representation) {
+                    $results[] = $resolver($representation);
+                }
+            }
```
Comment (olivernybroe) [inline src/Federation/Resolvers/Entities.php:54]: Could we think of a way to make the resolver itself responsible for this instead, so the if else statement wouldn't be needed? 🤔

**Review by olivernybroe — APPROVED** [review]
Looks pretty good 👍  

Just some small ideas to consider for improving the readability of the code.

File: src/Federation/Resolvers/Entities.php [hunk 2]
```diff
@@ -32,11 +33,25 @@ public function __invoke($root, array $args, GraphQLContext $context, ResolveInf
     {
         $results = [];
 
+        /** @var array<string, array<int, array<string, mixed>>> $groupedRepresentations */
+        $groupedRepresentations = [];
```
[the remark this answers is not in the document]
Comment (spawnia) [reply src/Federation/Resolvers/Entities.php:37]: Apart from `__typename`, there are no known properties, so this really just is a map of a list of maps 😅

File: src/Federation/Resolvers/Entities.php (hunk 1 above)
Comment (spawnia) [reply src/Federation/Resolvers/Entities.php:54]: This conditional maintains backwards compatibiltity with non-batched entity resolvers. I would not know how to solve this any other way, without forcing developers to change their code or write more then the bare minimum code required.

Also, by doing grouping and merging here, we could change the logic of ordering around without the developer having to do anything.

**Comment by spawnia** [comment]
@olivernybroe thanks for the review. Do you want to test the PR before I merge, or is the design good to go? We can change internals later, but I would like to have the API stabilized.

File: src/Federation/EntityResolverProvider.php [hunk 3]
```diff
@@ -135,9 +142,16 @@ protected function resolverFromClass(string $typename): ?Closure
             return null;
         }
 
+        if (is_a($resolverClass, BatchedEntityResolver::class, true)) {
+            return app($resolverClass);
```
[the remark this answers is not in the document]
Comment (olivernybroe) [reply src/Federation/EntityResolverProvider.php:146]: No, but you can still inject the Container itself, and then replace `app` helper with `$container->get($resolverClass)` 👍

[the remark this answers is not in the document]
Comment (spawnia) [reply src/Federation/EntityResolverProvider.php:146]: Good point, that is more efficient. Done

**Pull request merged**